### ISSUE TITLE
LDAP connector loses MFA status of user on subsequent login

### ISSUE THREAD

**Issue by jcpreston26** [FusionAuth/fusionauth-issues#1438]
### Issue

MFA status of a FusionAuth user that is not migrated during a connector authentication is lost.

### Steps to reproduce
1. Set up a tenant and link with an LDAP connector.
2. Set up an application in FusionAuth and link it with your web application.
3. User initially logs in, user gets created.
4. Configure user with MFA and/or application memberships (either through your web app workflow or manually via Fusion administrative pages)
5. Logout

(Updated by @mooreds) .

# Original issue

### Description 
Related to/Duplicate of #1220 and #1432 with an additional item.  The use of the LDAP Connector reconcile lambda is causing the user to not just lose application memberships and not have groups be updated, but also causing a loss of their multi-factor authentication (e.g. their TOTP/Authenticator app link).

### Affects versions
1.30.1

### Steps to reproduce
1. Set up a tenant and link with an LDAP connector.
2. Set up an application in FusionAuth and link it with your web application.
3. User initially logs in, user gets created.
4. Configure user with MFA and/or application memberships (either through your web app workflow or manually via Fusion administrative pages)
5. Logout
6. Have the user log in again.  The application membership and MFA information will no longer be available.

### Expected behavior
* A reconcile lambda would allow the system to either create a new user or modify an existing user if the source data provider has relevant modifications.  If there were a way for the lambda to lookup the existing user that could work. Or, if the lambda function was defined to have a parameter which was an existing user object matching the (tenant, email) tuple (or, null, if not found).
* As the multi-factor authentication is entirely handled within Fusion, there's no good way for that information to be pulled or managed from the LDAP source; the user wouldn't be given the TOTP QR code, for example.
* Once a user has configured multi-factor authentication, there should be no way to remove it without the user directly requesting it (or administrators being able to reset it after a review).  With the LDAP Connector, the MFA will be removed at every login.

### Platform
 - OS: CentOS 7
 - Database PostgresSQL 13

### Additional context
Unlike in #1220, use of the OIDC IdP is not an option for us as we are not using Azure AD.  

Our intended use case would be to allow our internal users to log in to the public applications using their Active Directory credentials for administrative actions.  As this may need to happen from the public internet, any routing to internal federation would not be a desirable option.  Also, the multi-factor authentication for a public login portal is not only our desired best practice but also a regulatory requirement.

I absolutely agree with the statement in the previous issues that "we didn't want to deal with merging data from another source of record, that can get really complicated."  Trying to figure that out to "just work" and cover all use cases would be awful.  A reconcile, however, would be a place to push those specific decisions off onto us as the deployers of the FusionAuth application.  We could make the choices specific to our needs as to which data to keep and which is irrelevant, allowing us to do the work to reconcile changes.

If an LDAP IdP would be easier or fit your internal models, that would also work.  The key issue currently is the loss of MFA when using an LDAP source.

**Comment by jcpreston26** [comment]
Yes.  Or, _a_ system of record.  My thought was: our internal users' passwords authenticating against our existing domain via LDAP, and our external (customer) users' passwords authenticating against the password database within Fusion (or, if requested, that customer's own accessible OIDC or SAML provider).

**Comment by jcpreston26** [comment]
Is the specific issue with MFA being reset going to have to wait on #1445?

**Comment by mooreds** [comment]
I'm not sure that #1445 will ever be implemented. Losing MFA status seems like a separate issue.

Sorry, looks like GH automatically closed this; it shouldn't have.

I'll re-open.

**Comment by robotdan** [comment]
@bguyza is this the same issue? https://github.com/FusionAuth/fusionauth-issues/issues/1432

**Comment by bguyza** [comment]
@robotdan 
I do not believe so unless MFA configuration were to be managed by the identity provider "source of record", or specifically, the LDAP server.

**Comment by robotdan** [comment]
We are debating this one internally. The connector is the source of record for this user, so we could fix this and preserve the 2FA configuration, but I tend to think it is working as designed. 

The connector is currently is charge of returning the source of record for the user until the user is migrated. This includes mostly everything about the user including registrations and group memberships. 

If we were to preserve 2FA configuration, this has a few implications:
1. You cannot mange 2FA for a user from a connector. 
2. You cannot migrate 2FA for a user from a connector. 
3. You cannot only manage 2FA for a user from FusionAuth, even when using a connector.  

And if we were to manage it, then 2FA becomes an exception to the rule. 

I tend to want to leave this as is - because it allows the most flexibility in the connector, and it affirms the connector to be the source of record. 

There is some nuance here as well, since the 2FA configuration is not necessarily an attribute of the user, but a configuration  for the user. I could probably convince myself of either solution.

Anyone have an opinion here? I suppose we could also add some additional configuration to the connector to let you decide, but this further complicates the configuration and the code.

**Comment by jcpreston26** [comment]
FWIW (with a caveat that I'm coming at it from my perspective and there's likely other approaches):
As a practical matter, I would assume that If I'm enabling 2FA in the Fusion application, it is because I'm intending to have Fusion manage the 2FA as the connector I'm using either can't or doesn't provide that functionality.

**Comment by robotdan** [comment]
Thanks for the feedback @jcpreston26. 

> As a practical matter, I would assume that If I'm enabling 2FA in the Fusion application, it is because I'm intending to have Fusion manage the 2FA as the connector I'm using either can't or doesn't provide that functionality.

I think this is a valid perspective. It is likely common that the connector will not provide this function. 

The tricky part is that we would either need to open up some additional config to allow you to configure this is your desired behavior, or we'd have to keep track of how owns the 2FA config in particular (FusionAuth or the current connector), and then have to think about what happens if both parties try to provide configuration. 

I suppose if we were to choose one of those two options, I would vote to add additional connector or perhaps tenant configuration to indicate that 2FA is managed by the FusionAuth Tenant or the connector assigned to the user. 

We'll discuss this one further. 

@voidmain any opinion here?

**Comment by voidmain** [comment]
The only issue I can see is how to distinguish between the lack of MFA support at the Connector and the Connector disabling MFA. We could store the MFA Connector separately from the authentication Connector and that might allow this to work as desired.

**Comment by robotdan** [comment]
Internal notes:

Likely adding a policy will allow us to solve this use case. The policy would allow a connector to indicate it is managing 2FA for a user. 

We could then optionally enforce this policy at the API level to fail 2FA modifications for a user that has 2FA managed by their connector policy.

**Comment by mooreds** [comment]
Related but different: https://github.com/FusionAuth/fusionauth-issues/issues/1821
